server/src/main/java/io/druid/server/listener/announcer/ListenerDiscoverer.java

```diff
@@ -0,0 +1,200 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.server.listener.announcer;
+
+import com.google.common.base.Predicate;
+import com.google.common.base.Strings;
+import com.google.common.base.Throwables;
+import com.google.common.collect.Collections2;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.net.HostAndPort;
+import com.google.inject.Inject;
+import com.metamx.common.lifecycle.LifecycleStart;
+import com.metamx.common.lifecycle.LifecycleStop;
+import com.metamx.common.logger.Logger;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.utils.ZKPaths;
+import org.apache.zookeeper.KeeperException;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+public class ListenerDiscoverer
+{
+  private static final Logger LOG = new Logger(ListenerDiscoverer.class);
+  private volatile Map<HostAndPort, Long> lastSeenMap = ImmutableMap.of();
+  private final CuratorFramework cf;
+  private final ListeningAnnouncerConfig listeningAnnouncerConfig;
+  private final Object startStopSync = new Object();
+  private volatile boolean started = false;
+
+  @Inject
+  public ListenerDiscoverer(
+      CuratorFramework cf,
+      ListeningAnnouncerConfig listeningAnnouncerConfig
+  )
+  {
+    this.cf = cf;
+    this.listeningAnnouncerConfig = listeningAnnouncerConfig;
+  }
+
+  @LifecycleStart
+  public void start()
+  {
+    synchronized (startStopSync) {
+      if (started) {
+        LOG.debug("Already started");
+        return;
+      }
+      started = true;
+      LOG.info("Started");
+    }
+  }
+
+  @LifecycleStop
+  public void stop()
+  {
+    synchronized (startStopSync) {
+      if (!started) {
+        LOG.debug("Already stopped");
+        return;
+      }
+      LOG.info("Stopped");
+      started = false;
+    }
+  }
+
+  /**
+   * Get nodes at a particular listener.
+   * This method lazily adds service discovery
+   *
+   * @param listener_key The Listener's service key
+   *
+   * @return A collection of druid nodes as established by the service discovery
+   *
+   * @throws IOException if there was an error refreshing the zookeeper cache
+   */
+  public Collection<HostAndPort> getNodes(final String listener_key) throws IOException
+  {
+    return getCurrentNodes(listener_key).keySet();
+  }
+
+  Map<HostAndPort, Long> getCurrentNodes(final String listener_key) throws IOException
+  {
+    final HashMap<HostAndPort, Long> retVal = new HashMap<>();
```

Comment:
I had considered that, but when I tried to implement it there was a lot of complexity and trouble, especially in unit tests and transient errors. So I went with a more straightforward CuratorFramework approach directly, especially since the utilization for CF is minimal.

Eventually the CF here should be replaced by a more general announcement framework.